File: admin/src/api/voice.ts
import client, { unwrap } from './client';
import type { ApiResponse, VoiceProfile } from '@zhiqu/shared';

export function listVoiceProfiles(includeInactive = true) {
  return unwrap<VoiceProfile[]>(
    client.get('/admin/voice/profiles', { params: { include_inactive: includeInactive } }),
  );
}

export function createVoiceProfile(data: {
  name: string;
  description?: string;
  provider: 'tts' | 'openvoice';
  voice_key?: string;
  is_active: boolean;
  sort_order: number;
  file?: File;
}) {
  const form = new FormData();
  form.append('name', data.name);
  if (data.description) form.append('description', data.description);
  form.append('provider', data.provider);
  if (data.voice_key) form.append('voice_key', data.voice_key);
  form.append('is_active', String(data.is_active));
  form.append('sort_order', String(data.sort_order));
  if (data.file) form.append('file', data.file);
  return unwrap<VoiceProfile>(
    client.post('/admin/voice/profiles', form, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 120_000,
    }),
  );
}

export async function transcribeAudio(file: Blob): Promise<{ text: string }> {
  const form = new FormData();
  form.append('file', file, 'recording.webm');
  const res = await client.post<ApiResponse<{ text: string }>>('/admin/voice/stt', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 120_000,
  });
  return res.data.data;
}

export async function synthesizeSpeech(text: string, voiceProfileId?: string | null): Promise<Blob> {
  const res = await client.post('/admin/voice/tts', {
    text,
    voice_profile_id: voiceProfileId || null,
  }, {
    responseType: 'blob',
    timeout: 120_000,
  });
  return res.data as Blob;
}

File: services/interfaces/api/admin/voice.py
"""Admin voice profile management APIs."""

from __future__ import annotations

from fastapi import APIRouter, File, Form, UploadFile
from fastapi.responses import Response

from interfaces.api.deps import AdminUser, VoiceSvc
from interfaces.schemas.voice import SpeechRequest, VoiceProfileOut
from shared.response import ok

router = APIRouter(prefix="/api/v1/admin/voice", tags=["admin-voice"])


@router.get("/profiles", summary="音色列表")
async def list_profiles(svc: VoiceSvc, _admin: AdminUser, include_inactive: bool = True):
    rows = await svc.list_profiles(include_inactive=include_inactive)
    return ok([VoiceProfileOut.model_validate(r).model_dump(mode="json") for r in rows])


@router.post("/profiles", summary="上传或登记音色")
async def create_profile(
    svc: VoiceSvc,
    admin: AdminUser,
    name: str = Form(...),
    description: str | None = Form(None),
    provider: str = Form("tts"),
    voice_key: str | None = Form(None),
    is_active: bool = Form(True),
    sort_order: int = Form(0),
    file: UploadFile | None = File(None),
):
    audio = await file.read() if file else None
    row = await svc.create_profile(
        name=name,
        description=description,
        provider=provider,
        voice_key=voice_key,
        is_active=is_active,
        sort_order=sort_order,
        created_by=admin.sub,
        reference_filename=file.filename if file else None,
        reference_content_type=file.content_type if file else None,
        reference_audio=audio,
    )
    return ok(VoiceProfileOut.model_validate(row).model_dump(mode="json"))


@router.post("/stt", summary="语音转文本")
async def transcribe(_admin: AdminUser, svc: VoiceSvc, file: UploadFile = File(...)):
    audio = await file.read()
    text = await svc.transcribe(
        filename=file.filename or "recording.wav",
        content_type=file.content_type or "audio/wav",
        audio=audio,
    )
    return ok({"text": text})


@router.post("/tts", summary="文本转语音")
async def synthesize(body: SpeechRequest, _admin: AdminUser, svc: VoiceSvc):
    audio, content_type = await svc.synthesize(
        text=body.text,
        profile_id=body.voice_profile_id,
    )
    return Response(content=audio, media_type=content_type)
